Add a Semantic Convention for Serverless workload (#8)
dynatrace-oss-contrib/opentelemetry-specification: specification/data-faas.md

> # Semantic conventions for FaaS spans

Reviewer: Can we say something to the effect of "If Spans following this convention are produced, a Resource of type `faas` MUST exist." (with link to resource conventions)."?

Maybe (I would like that better) put the resource conventions right in this document, instead of data-resource-semantic-conventions, put a link to the faas resource into data-resource-semantic-conventions.

Author: I would argue to use the former option. In my opinion, it is better to follow the existing separation where resources are defined in a document and span attributes in another.

Reviewer: Well for now it's totally fine, but in future, when the resource conventions document gets too long and we split it up, it would make more sense to mee to have a single `data-faas` file instead of `data-faas` for spans and `data-resource-faas` for resources.

This document defines how to describe an instance of a function that runs without provisioning or managing of servers (also known as serverless) with spans.

> Span `name` should be set to function name being executed. Depending on the trigger of the function, additional attributes MUST be set. For example, an HTTP trigger must set the `SpanKind` to `Server` and follow the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions).

Reviewer 2: ```suggestion
Span `name` should be set to the function name being executed. Depending on the trigger of the function, additional attributes MUST be set. For example, an HTTP trigger must set the `SpanKind` to `Server` and follow the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions).
```

# Semantic conventions for FaaS spans

This document defines how to describe an instance of a function that runs without provisioning or managing of servers (also known as serverless) with spans.

> Span `name` should be set to the function name being executed. Depending on the trigger of the function, additional attributes MUST be set. For example, an HTTP trigger MUST follow the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions).

Reviewer 3: please avoid that long lines.

Reviewer 4: I actually prefer long markdown lines. It means my editor can deal with line breaks how I have it configured, and the `suggestion` feature of github works much better if you only have to change a single line.

> Span `name` should be set to the function name being executed. Depending on the trigger of the function, additional attributes MUST be set. For example, an HTTP trigger MUST follow the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions).

Reviewer 3: I'm not sure if it is always possible to follow e.g. HTTP Server semantic conventions. For example an AWS lambda function may be triggered using AWS SDK which in turn uses HTTP.
Or AWS Lambda function triggered by AWS SQS. Function is triggered once but may get a bulk of messages. One would expect to see one span per message (linked to the corresponding sender) and not one per lambda function call (not linkable to anyone).

Author: It is indeed true that the SDK internally uses an HTTP call to trigger a lambda. However, the lambda itself might not be configured to process an HTTP request. This attribute is not for how a lambda is called but more on the type of event the lambda is configured to answer to.

Reviewer 3: Ah ok. maybe write `Depending on the function input data` or similar?

Author: I understand the confusion. I agree, the wording could be improved

If Spans following this convention are produced, a Resource of type `faas` MUST exist following the [Resource semantic convention](data-resource-semantic-conventions.md#function-as-a-service). 

| Attribute name  | Notes  and examples  | Required? |
|---|---|--|
| `component` | Denotes the type of the span and needs to be `"faas"` | Yes | 
| `faas.trigger` | Type of the trigger on which the function is executed. <br > The following spelling SHOULD be used for trigger strings: "event", "http", "manual", "pubsub", "timer". | Yes |

> | `faas.id` | String containing the unique execution id of the function. E.g. `af9d5aa4-a685-4c5f-a22b-444f80b3cc28` | No |

Reviewer 4: Propose renaming `faas.id` to `faas.execution`. `id` is too general a term and is too easily misread and misunderstood.

In addition, you already have `faas.identifier`. Having `id` which is typically short for `identifier` mean something different than `identifier` is asking for trouble.

## Difference between id and instance

For performance reasons (e.g. [AWS lambda], or [Azure functions]), FaaS providers allocate an execution environment for a single instance of a function that is used to serve multiple requests.

> Developers exploit this fact to solve the **cold start** issue, caching expensive resources computation between different function execution. 

Reviewer 4: ```suggestion
Developers exploit this fact to solve the **cold start** issue, caching expensive resource computations between different function executions. 
```

> Furthermore, FaaS providers encourage this behavior, e.g. [Google functions].

Reviewer 4: Do you have a link for this? Without one, this sentence should probably just be removed.

Author: Yes, it is in "[Google functions]", which is defined at the end of the file. In the rendering it will be automatically linked there.

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [General Attributes](#general-attributes)
  * [Difference between execution and instance](#difference-between-execution-and-instance)
- [Function Type](#function-type)
  * [Datasource](#datasource)
  * [HTTP](#http)
  * [PubSub](#pubsub)
  * [Timer](#timer)
  * [Other](#other)

<!-- tocstop -->

# General Attributes

Span `name` should be set to the function name being executed. Depending on the value of the `faas.trigger` attribute, additional attributes MUST be set. For example, an `http` trigger MUST follow the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions). For more information, refer to the [Function Type](#function-type) section.

If Spans following this convention are produced, a Resource of type `faas` MUST exist following the [Resource semantic convention](data-resource-semantic-conventions.md#function-as-a-service). 

| Attribute name  | Notes  and examples  | Required? |
|---|---|--|
| `component` | Denotes the type of the span and needs to be `"faas"` | Yes | 

> | `faas.trigger` | Type of the trigger on which the function is executed. <br > it SHOULD be one of the following strings: "datasource", "http", "pubsub", "timer", or "other". | Yes |

Reviewer 4: nit

```suggestion
| `faas.trigger` | Type of the trigger on which the function is executed. <br > It SHOULD be one of the following strings: "datasource", "http", "pubsub", "timer", or "other". | Yes |
```

> | `faas.trigger` | Type of the trigger on which the function is executed. <br > The following spelling SHOULD be used for trigger strings: "datasource", "http", "pubsub", "timer", or "other". | Yes |

Reviewer 4: It's a little odd to specify this as a spelling concern. I would do something more like
```suggestion
| `faas.trigger` | Type of the trigger on which the function is executed. <br > it SHOULD be one of the following strings: "datasource", "http", "pubsub", "timer", or "other". | Yes |
```

## Difference between execution and instance

For performance reasons (e.g. [AWS lambda], or [Azure functions]), FaaS providers allocate an execution environment for a single instance of a function that is used to serve multiple requests.
Developers exploit this fact to solve the **cold start** issue, caching expensive resource computations between different function executions. 
Furthermore, FaaS providers encourage this behavior, e.g. [Google functions].
Therefore, the span attribute `faas.execution` differs from the resource attribute `faas.instance` in the following:

- `faas.execution` refers to the current request ID handled by the function;
- `faas.instance` refers to the execution environment ID of the function.

[AWS lambda]: https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html
[Azure functions]: https://docs.microsoft.com/en-us/azure/azure-functions/manage-connections#static-clients
[Google functions]: https://cloud.google.com/functions/docs/concepts/exec#function_scope_versus_global_scope

# Function Type

This section describes how to handle the span creation and additional attributes based on the value of the attribute `faas.trigger`.

## Datasource

> The function is meant to answer to a single event generated by a datasource, e.g. DynamoDB, Cloud Storage, Cosmos DB.

Reviewer 3: Is the lambda function really serving a database request in this case? I have never used this so I'm not sure.

Reviewer 4: You can configure lambda to be triggered when a row is inserted into a DB, sort of like a stored procedure. It does not serve the request, just may handle some business logic like updating a cache or sending an email or something.

example use case: If a row is inserted into the `PasswordResetTokens` table, send the user an email with a password reset link.

This document defines how to describe an instance of a function that runs without provisioning or managing of servers (also known as serverless) with spans.

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [General Attributes](#general-attributes)
  * [Difference between execution and instance](#difference-between-execution-and-instance)
- [Function Type](#function-type)
  * [Datasource](#datasource)
  * [HTTP](#http)
  * [PubSub](#pubsub)
  * [Timer](#timer)
  * [Other](#other)

<!-- tocstop -->

# General Attributes

Span `name` should be set to the function name being executed. Depending on the value of the `faas.trigger` attribute, additional attributes MUST be set. For example, an `http` trigger MUST follow the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions). For more information, refer to the [Function Type](#function-type) section.

If Spans following this convention are produced, a Resource of type `faas` MUST exist following the [Resource semantic convention](data-resource-semantic-conventions.md#function-as-a-service). 

| Attribute name  | Notes  and examples  | Required? |
|---|---|--|

> | `faas.trigger` | Type of the trigger on which the function is executed. <br > It SHOULD be one of the following strings: "datasource", "http", "pubsub", "timer", or "other". | Yes |

Reviewer 5: w/o knowing too much about OTel policy: 
my gut feeling is, that this generic trigger definition attempts to unify very specific entities of various faas service providers. The benefit of this unification is yet unclear. 
I would consider to introduce a faas service provider (e.g. AWS, GCF etc.) and specific triggers for these service providers.

This section describes how to handle the span creation and additional attributes based on the value of the attribute `faas.trigger`.

## Datasource

A datasource function is triggered as a response to some data source operation such as a database or filesystem read/write.
For `faas` spans with trigger `datasource`, it is recommended to set the following attributes. 

| Attribute name  | Notes  and examples  | Required? |
|---|---|--|
| `faas.document.collection` | The name of the source on which the operation was perfomed. For example, in Cloud Storage or S3 corresponds to the bucket name, and in Cosmos DB to the database name. | Yes |
| `faas.document.operation`  | Describes the type of the operation that was performed on the data.<br /> It SHOULD be one of the following strings: "insert", "edit", "delete". | Yes |
| `faas.document.time`       | A string containing the time when the data was accessed in the [ISO 8601] format expressed in [UTC]. E.g. `"2020-01-23T13:47:06Z"` | Yes |
| `faas.document.name`       | The document name/table subjected to the operation.<br /> For example, in Cloud Storage or S3 is the name of the file, and in Cosmos DB the table name.  | No |

## HTTP

> The function responsibility is to provide an answer to an inbound HTTP request. The `faas` span MUST follow the recomandations described in the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions).

Reviewer 5: In case AWS-SDK invoke API falls into http trigger category, it cannot follow any HTTP semantic conventions

This document defines how to describe an instance of a function that runs without provisioning or managing of servers (also known as serverless) with spans.

> Span `name` should be set to the function name being executed. Depending on the trigger of the function, additional attributes MUST be set. For example, an HTTP trigger MUST set the `SpanKind` to `Server` and MUST follow the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions).

Author: Right!

This document defines how to describe an instance of a function that runs without provisioning or managing of servers (also known as serverless) with spans.

> Span `name` should be set to the function name being executed. Depending on the value of the `faas.trigger` attribute, additional attributes MUST be set. For example, an `http` trigger MUST follow the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions).

Reviewer: But is that really doable? I assume in many cases the HTTP span would be created by another sensor that comes before or after the faas sensor. Is the faas sensors supposed to integrate into the other sensor or is it fine if we have nested spans instead of all the information on a single span?

Author: In my understanding, a FaaS sensor decides the correct `faas.trigger` value and then forwards  accordingly the tagging to the correct sensor

Reviewer 3: At least for AWS lambda NodeJs the translation from HTTP to lambda function call (incl. load balancing) happens outside of the processes hosting the function. There is no real HTTP happening inside lambda, a function is invoked because of an HTTP request and the function gets input data converted into an event format (see e.g. https://docs.aws.amazon.com/lambda/latest/dg/with-on-demand-https.html and https://docs.aws.amazon.com/lambda/latest/dg/lambda-services.html#intro-core-components-event-sources).

Reviewer 4: I think it makes sense to keep this language. In most cases there will be no http span before the faas span anyways, and in the rare (nonexistent?) case that there are two nested spans, it is ok I think to duplicate some information.

Reviewer: I think adding this requirements could cause headaches. E.g. if a sensors only offers an enable/disable API but no "add attributes from *this* request object to *this* span" API, then implementing this will be hard. Also, this would mean that faas depends on HTTP, messaging, ... sensors.

Author: I share your concerns, however I think this requirement should be in. Without such attributes, a possible correlation on the source that triggered the lambda is not possible. 
My opinion is also given by the fact that lambda is anyway a special use case for OpenTelemetry and with such, the faas sensor should be handle as a special case.

Reviewer 4: How does this add a dependency on another sensor? In most cases, the faas sensor will be the local root. it will just look in the request object, notice it is an http style call, and add the http attributes in addition to the faas attributes. I would expect the http sensor not to even create an incoming span most of the time.

Reviewer: Ah, so you mean since HTTP-style faas invocations won't go through a normal HTTP code path, a normal HTTP sensors is not required since the faas needs to re-implement a faas HTTP sensor anyway. Makes sense! 👍  Having understood that, I'm fine with this requirement. 😃

Reviewer 3: > I think it makes sense to keep this language. In most cases there will be no http span before the faas span anyways, and in the rare (nonexistent?) case that there are two nested spans, it is ok I think to duplicate some information

Having a root span just for FaaS may cause issues in e.g. AWS SQS where a bulk of messages is passed to the function. Each message may contain it's own tag and therefore we should end up in one Span per message correctly linked to the remote sender. If there is a root span these message spans would use this one as parent and the link to the senders may be lost.

Main question here is what should our service detection show? One service per function invocation or more concrete what the function is used for.

Reviewer 4: In that case would the sensor just not create a `faas` root span and instead create one `faas` root span for each message, using the `messaging` semconv?

Reviewer 3: It depends on the implementation - which is done based on the spec.

Author: I agree with @Flarna that we need to specify better each case. At the moment, we don't have a semconv for messaging queue (in the otel-spec repository). I suggest to add a TODO and solve it based on the feedback received from when we PR https://github.com/dynatrace-oss-contrib/opentelemetry-specification/pull/1

> Span `name` should be set to the function name being executed. Depending on the value of the `faas.trigger` attribute, additional attributes MUST be set. For example, an `http` trigger MUST follow the [HTTP Server semantic conventions](data-http.md#http-server-semantic-conventions).

Author: Does it look better now? @Flarna

This section describes how to handle the span creation and additional attributes based on the value of the attribute `faas.trigger`.

## Datasource

> A function is set to respond as callback when operations on a datasource are perfomed.

Reviewer 4: ```suggestion
A datasource function is triggered as a response to some data source operation such as a database or filesystem read/write.
```
